### PR TITLE
[Document Loader] Add AzureAIContentUnderstandingLoader document loader

### PR DESCRIPTION
**Pull request by chienyuanchang**
### Summary

Adds a new LangChain document loader that integrates with [Azure Content Understanding](https://learn.microsoft.com/azure/ai-services/content-understanding/overview) — a multimodal AI service for extracting semantic content from documents, images, audio, and video.

### New class: `AzureAIContentUnderstandingLoader`

A `BaseLoader` subclass that accepts a local file, URL, or raw bytes and yields LangChain `Document` objects with extracted markdown content and rich metadata (fields, confidence scores, source info).

**Key features:**
- **Multimodal support** — PDF, images, audio, and video via a single loader class
- **Three output modes:**
  - `markdown` (default) — one Document per content item with full markdown and extracted fields
  - `page` — one Document per page (document content only; falls back to markdown for audio/video)
  - `segment` — one Document per content segment with per-segment fields and category. Works with any content kind when the analyzer has `enableSegment=true` (e.g., topic-based video segments, document classification segments)
- **Auto-selects prebuilt analyzer** based on MIME type; also supports custom analyzers with field extraction, classification, and model deployments
- **Sync and async** — `lazy_load()` / `alazy_load()` with native async CU client
- **Content range** — analyze subsets (e.g., pages 1-3, first 60 seconds of audio)
- **`analyze_kwargs`** — forward extra parameters to the CU API (e.g., `processing_location`)

### Files changed

| File | Description |
|------|-------------|
| `langchain_azure_ai/document_loaders/__init__.py` | Lazy-import module exposing `AzureAIContentUnderstandingLoader` and `OutputMode` |
| `langchain_azure_ai/document_loaders/content_understanding.py` | Loader implementation|
| `langchain_azure_ai/document_loaders/_constants.py` | MIME aliases, media-type-to-analyzer map, default analyzer constant |
| `docs/content_understanding_loader_demo.ipynb` | Demo notebook covering all features |
| `tests/unit_tests/test_content_understanding_loader.py` | 100+ unit tests (init validation, MIME detection, all output modes, field flattening, edge cases) |
| `tests/integration_tests/test_content_understanding_loader.py` | 16 integration tests — 14 sync with VCR cassettes for offline replay, 2 async live-only |
| `tests/integration_tests/cassettes/*.yaml` | 14 recorded HTTP cassettes (sanitized, no credentials) |
| `pyproject.toml` | Added `azure-ai-contentunderstanding` dep, mypy overrides, `[contentunderstanding]` extra |
| `poetry.lock` / `uv.lock` | Lock files updated |
| `README.md` | Minor update |

### Dependencies
- `azure-ai-contentunderstanding >=1.0.0, <2.0.0` (new runtime dependency)
- `filetype >=1.2, <2` — optional extra (`pip install langchain-azure-ai[contentunderstanding]`) for binary MIME sniffing
- `vcrpy >=7.0` and `pytest-recording ~=0.13.4` (test-only, already in dev deps)

### CODE REVIEW

File: libs/azure-ai/langchain_azure_ai/document_loaders/content_understanding.py [hunk 1]
```diff
@@ -0,0 +1,541 @@
...
+
+_DEFAULT_ANALYZERS: Dict[str, str] = {
+    "application/": "prebuilt-documentSearch",
+    "text/": "prebuilt-documentSearch",
+    "image/": "prebuilt-documentSearch",
+    "audio/": "prebuilt-audioSearch",
+    "video/": "prebuilt-videoSearch",
+}
+
+_VALID_OUTPUT_MODES = ("markdown", "page", "segment")
```
Comment (yungshinlintw) [inline libs/azure-ai/langchain_azure_ai/document_loaders/content_understanding.py:52]: Consider creating an enum and then put that enum into __init__.py: `__all__`

File: libs/azure-ai/langchain_azure_ai/document_loaders/content_understanding.py [hunk 2]
```diff
@@ -0,0 +1,541 @@
...
+                ``AzureKeyCredential``, or ``TokenCredential``.
+            analyzer_id: Analyzer to use. Defaults by input MIME type if omitted.
+            file_path: Path to a local file.
+            url: URL to the content.
+            bytes_source: Raw bytes of the content.
+            source: Label for ``metadata["source"]``. Defaults to *file_path*
+                or *url* when provided.
+            output_mode: How to split results into Documents —
+                ``"markdown"`` (default), ``"page"``, or ``"segment"``.
+            content_range: Subset of input to analyze. Pages use ``"1-3,5,9-"``;
```
Comment (yungshinlintw) [inline libs/azure-ai/langchain_azure_ai/document_loaders/content_understanding.py]: Suggestion: Pages use 1-based page number, such as ...

File: libs/azure-ai/langchain_azure_ai/document_loaders/content_understanding.py [hunk 3]
```diff
@@ -0,0 +1,541 @@
...
+        credential: Union[str, AzureKeyCredential, TokenCredential],
+        *,
+        analyzer_id: Optional[str] = None,
+        file_path: Optional[str] = None,
+        url: Optional[str] = None,
+        bytes_source: Optional[bytes] = None,
+        source: Optional[str] = None,
+        output_mode: str = "markdown",
+        content_range: Optional[str] = None,
+        output_selection: Optional[List[str]] = None,
```
Comment (yungshinlintw) [inline libs/azure-ai/langchain_azure_ai/document_loaders/content_understanding.py]: Given that this is for selecting meta data, mabye `metadata_selection`?

File: libs/azure-ai/langchain_azure_ai/document_loaders/content_understanding.py [hunk 4]
```diff
@@ -0,0 +1,541 @@
...
+                or *url* when provided.
+            output_mode: How to split results into Documents —
+                ``"markdown"`` (default), ``"page"``, or ``"segment"``.
+            content_range: Subset of input to analyze. Pages use ``"1-3,5,9-"``;
+                audio/video uses milliseconds ``"0-60000"``.
+            output_selection: What to include in metadata, e.g.
+                ``["fields", "tables"]``. Defaults to include fields.
+            model_deployments: Mapping of model names to deployment names.
+                Required for custom analyzers that use ``generate`` or
+                ``classify`` field methods.
```
Comment (yungshinlintw) [inline libs/azure-ai/langchain_azure_ai/document_loaders/content_understanding.py]: Why is this required for custom analyzers? This is optional for users to override default model deployments

Comment (chienyuanchang) [reply libs/azure-ai/langchain_azure_ai/document_loaders/content_understanding.py]: Thank you for catching this. I will revise the docstring.

File: libs/azure-ai/langchain_azure_ai/document_loaders/content_understanding.py [hunk 5]
```diff
@@ -0,0 +1,541 @@
...
+        Object and array types are recursively flattened.
+        """
+        field_type = field.type
+
+        # Array fields → list of {value, confidence}
+        if field_type == "array" and field.value is not None:
+            return [
+                {
+                    "value": self._resolve_field_value(item),
+                    "confidence": item.confidence,
```
Comment (yungshinlintw) [inline libs/azure-ai/langchain_azure_ai/document_loaders/content_understanding.py]: If `field` has no confidence, we will throw here.

Comment (yungshinlintw) [reply libs/azure-ai/langchain_azure_ai/document_loaders/content_understanding.py]: Same as other field.confidence statement below

File: libs/azure-ai/pyproject.toml
Comment (yungshinlintw) [inline libs/azure-ai/pyproject.toml:3]: We need to change this value for a new version?

File: libs/azure-ai/tests/unit_tests/test_content_understanding_loader.py [hunk 6]
```diff
@@ -0,0 +1,1937 @@
...
+
+
+def _make_field(
+    field_type: str,
+    confidence: float = 0.95,
+    **kwargs: Any,
+) -> Mock:
+    field = Mock()
+    field.type = field_type
+    field.confidence = confidence
```
Comment (yungshinlintw) [inline libs/azure-ai/tests/unit_tests/test_content_understanding_loader.py:47]: We will have cases that a field has no confidence.

File: libs/azure-ai/langchain_azure_ai/document_loaders/content_understanding.py [hunk 7]
```diff
@@ -0,0 +1,541 @@
...
+            return mime_type
+        return None
+
+    def _resolve_default_analyzer(self) -> str:
+        """Pick a default analyzer based on MIME type prefix."""
+        if self._mime_type:
+            for prefix, analyzer in _DEFAULT_ANALYZERS.items():
+                if self._mime_type.startswith(prefix):
+                    return analyzer
+        return "prebuilt-documentSearch"
```
Comment (yungshinlintw) [inline libs/azure-ai/langchain_azure_ai/document_loaders/content_understanding.py]: Given that the user can use application/octet-stream, CU MAF integration does a mime type sniffing if the bytes are available. We can consider alignment here.

File: libs/azure-ai/langchain_azure_ai/document_loaders/content_understanding.py [hunk 8]
```diff
@@ -0,0 +1,541 @@
...
+        )
+
+        try:
+            analysis_input = self._build_analysis_input()
+            poller = client.begin_analyze(
+                analyzer_id=self._analyzer_id,
+                inputs=[analysis_input],
+                model_deployments=self._model_deployments,
+                **self._analyze_kwargs,
+            )
```
Comment (yungshinlintw) [inline libs/azure-ai/langchain_azure_ai/document_loaders/content_understanding.py]: For bytes data, we should use begin_analyze_binary, as base64 (used in begin_analyze) will add 33% overhead

File: libs/azure-ai/pyproject.toml
Comment (chienyuanchang) [reply libs/azure-ai/pyproject.toml:3]: We will need to check with Facundo about his release plan and processure.

File: libs/azure-ai/langchain_azure_ai/document_loaders/content_understanding.py (hunk 1 above)
Comment (chienyuanchang) [reply libs/azure-ai/langchain_azure_ai/document_loaders/content_understanding.py:52]: I added enum. Could you please help check if the implementation is what you expected?

File: libs/azure-ai/langchain_azure_ai/document_loaders/content_understanding.py (hunk 2 above)
Comment (chienyuanchang) [reply libs/azure-ai/langchain_azure_ai/document_loaders/content_understanding.py]: Revised

File: libs/azure-ai/langchain_azure_ai/document_loaders/content_understanding.py (hunk 3 above)
Comment (chienyuanchang) [reply libs/azure-ai/langchain_azure_ai/document_loaders/content_understanding.py]: Yes, revised

File: libs/azure-ai/langchain_azure_ai/document_loaders/content_understanding.py (hunk 4 above)
Comment (chienyuanchang) [reply libs/azure-ai/langchain_azure_ai/document_loaders/content_understanding.py]: Revised

File: libs/azure-ai/langchain_azure_ai/document_loaders/content_understanding.py (hunk 5 above)
Comment (chienyuanchang) [reply libs/azure-ai/langchain_azure_ai/document_loaders/content_understanding.py]: Added handling for no confidence

File: libs/azure-ai/tests/unit_tests/test_content_understanding_loader.py (hunk 6 above)
Comment (chienyuanchang) [reply libs/azure-ai/tests/unit_tests/test_content_understanding_loader.py:47]: Added test cases without confidence

File: libs/azure-ai/langchain_azure_ai/document_loaders/content_understanding.py (hunk 8 above)
Comment (chienyuanchang) [reply libs/azure-ai/langchain_azure_ai/document_loaders/content_understanding.py]: Added detection to use begin_analyze_binary

File: libs/azure-ai/langchain_azure_ai/document_loaders/content_understanding.py (hunk 7 above)
Comment (chienyuanchang) [reply libs/azure-ai/langchain_azure_ai/document_loaders/content_understanding.py]: Added functions to use sniffing when extension is missing or unhelpful. However, the `filetype` package is optional now. Need to do `pip install langchain-azure-ai[contentunderstanding]` to enable this feature.

File: libs/azure-ai/pyproject.toml
Comment (chienyuanchang) [reply libs/azure-ai/pyproject.toml:3]: @santiagxf will help with the version and release. Thank you!

File: libs/azure-ai/langchain_azure_ai/document_loaders/content_understanding.py [hunk 9]
```diff
@@ -0,0 +1,693 @@
...
+
+from langchain_azure_ai.document_loaders._constants import (
+    DEFAULT_ANALYZER,
+    MEDIA_TYPE_ANALYZER_MAP,
+    MIME_ALIASES,
+)
+
+logger = logging.getLogger(__name__)
+
+_USER_AGENT = "langchain-azure-ai-cu-loader/1.0.0"
```
Comment (santiagxf) [inline libs/azure-ai/langchain_azure_ai/document_loaders/content_understanding.py]: Can we use `langchain-azure-ai` only? That's the one we use in the rest of the repository. It should append to the existing user agent chain.

Comment (chienyuanchang) [reply libs/azure-ai/langchain_azure_ai/document_loaders/content_understanding.py]: Yes, will change. Thank you!

File: libs/azure-ai/langchain_azure_ai/document_loaders/content_understanding.py [hunk 10]
```diff
@@ -0,0 +1,693 @@
...
+                credential=DefaultAzureCredential(),
+                file_path="report.pdf",
+            )
+            docs = loader.load()
+    """
+
+    def __init__(
+        self,
+        endpoint: str,
+        credential: Union[str, AzureKeyCredential, TokenCredential],
```
Comment (santiagxf) [inline libs/azure-ai/langchain_azure_ai/document_loaders/content_understanding.py]: Is content understanding available in Foundry projects? If so, we should include `project_endpoint` option too. This will get the base endpount URL from the project. Most AI Services are available here. Not sure about content understanding though, but for example Document Intelligence is.

File: libs/azure-ai/langchain_azure_ai/document_loaders/content_understanding.py [hunk 11]
```diff
@@ -0,0 +1,693 @@
...
+        *,
+        analyzer_id: Optional[str] = None,
+        file_path: Optional[str] = None,
+        url: Optional[str] = None,
+        bytes_source: Optional[bytes] = None,
```
Comment (santiagxf) [inline libs/azure-ai/langchain_azure_ai/document_loaders/content_understanding.py:96]: Looks these parameters are mutually exclusive, correct? Would be good to name it similarly? like `source_path`, `source_url`, and `source_bytes`? Just an opinion, please ignore if you think it's not important. I see other parsers have the same pattern you suggested so it's not that important. At least let's document it in the example.

File: libs/azure-ai/langchain_azure_ai/document_loaders/content_understanding.py (hunk 10 above)
Comment (chienyuanchang) [reply libs/azure-ai/langchain_azure_ai/document_loaders/content_understanding.py]: Thank you for pointing this out. Yes, CU uses AI service endpoint `https://my-resource.services.ai.azure.com`. I added logic to accept `project_endpoint` such as `https://my-resource.services.ai.azure.com/api/projects/my-project` and trim it as endpoint to be consistent with other classes.

File: libs/azure-ai/langchain_azure_ai/document_loaders/content_understanding.py (hunk 9 above)
Comment (chienyuanchang) [reply libs/azure-ai/langchain_azure_ai/document_loaders/content_understanding.py]: Revised to `langchain-azure-ai` only

File: libs/azure-ai/langchain_azure_ai/document_loaders/content_understanding.py (hunk 11 above)
Comment (chienyuanchang) [reply libs/azure-ai/langchain_azure_ai/document_loaders/content_understanding.py:96]: It's a great suggestion. However, most of LangChain loaders are using `file_path`. We could consider keeping it. I enhanced the descriptions of these three parameters first.

File: libs/azure-ai/langchain_azure_ai/document_loaders/content_understanding.py (hunk 1 above)
Comment (chienyuanchang) [reply libs/azure-ai/langchain_azure_ai/document_loaders/content_understanding.py:52]: Using `Literal` for the OutputMode

File: libs/azure-ai/docs/content_understanding_loader_demo.ipynb [hunk 12]
```diff
@@ -0,0 +1,928 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "b7c8d9e0",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Option 1: Load from a URL (public or SAS-signed)\n",
```
Comment (yungshinlintw) [inline libs/azure-ai/docs/content_understanding_loader_demo.ipynb:102]: We should mention that analyzer_id is not specified here, so we will pick one of the prebuilt-*Search based on modalities.

File: libs/azure-ai/docs/content_understanding_loader_demo.ipynb [hunk 13]
```diff
@@ -0,0 +1,928 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "c1d2e3f4",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Examine extracted fields with confidence scores\n",
```
Comment (yungshinlintw) [inline libs/azure-ai/docs/content_understanding_loader_demo.ipynb]: I think that we will only have a summary field in this case. So we can mention that too.

File: libs/azure-ai/docs/content_understanding_loader_demo.ipynb [hunk 14]
```diff
@@ -0,0 +1,928 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "a5b6c7d8",
+   "metadata": {},
+   "source": [
+    "### Field Extraction with Prebuilt Analyzers\n",
+    "\n",
+    "Use a domain-specific prebuilt analyzer (like `prebuilt-invoice`) to extract structured fields with confidence scores. Domain-specific analyzers require `model_deployments` to specify which model deployment to use.\n",
```
Comment (yungshinlintw) [inline libs/azure-ai/docs/content_understanding_loader_demo.ipynb]: Provide pointers to prebuilt analyzers https://learn.microsoft.com/azure/ai-services/content-understanding/concepts/prebuilt-analyzers

**Review by yungshinlintw — APPROVED** [review]

File: libs/azure-ai/docs/content_understanding_loader_demo.ipynb (hunk 12 above)
Comment (chienyuanchang) [reply libs/azure-ai/docs/content_understanding_loader_demo.ipynb:102]: Added, Thank you!

File: libs/azure-ai/docs/content_understanding_loader_demo.ipynb (hunk 13 above)
Comment (chienyuanchang) [reply libs/azure-ai/docs/content_understanding_loader_demo.ipynb]: Added the note.

File: libs/azure-ai/docs/content_understanding_loader_demo.ipynb (hunk 14 above)
Comment (chienyuanchang) [reply libs/azure-ai/docs/content_understanding_loader_demo.ipynb]: Added the link. Thank you!
